### PR TITLE
[PRMP-786] Edge CloudFront

### PR DESCRIPTION
**Pull request by RioKnightleyNHS**

### CODE REVIEW

File: .github/workflows/terraform-destroy-environment-manual.yml [hunk 1]
```diff
@@ -52,46 +52,63 @@ jobs:
...
+          ./venv/bin/pip3 install --upgrade pip
+          ./venv/bin/pip3 install boto3==1.33.11
 
-    - name: Terraform Set Workspace
-      id: workspace
-      run:  terraform workspace select ${{ github.event.inputs.sandbox_workspace}}
-      working-directory: ./infrastructure
-      shell: bash
+      # Run script to remove Lambda@Edge associations
+      - name: Remove Lambda@Edge & CloudFront associations
```
Comment (abbas-khan10) [inline .github/workflows/terraform-destroy-environment-manual.yml:105]: Might be worth extracting out into it's own job (called `remove_edge_associations`) similar to the cleanup versions, just so you can put a dependency on this being successful before trying to destroy, and on line 52 update to `needs: [cleanup_versions_process, remove_edge_associations]`

File: infrastructure/dynamo_db.tf [hunk 2]
```diff
@@ -40,6 +40,27 @@ module "document_reference_dynamodb_table" {
...
+  stream_enabled                 = false
+  ttl_enabled                    = true
+  ttl_attribute_name             = "TTL"
+  point_in_time_recovery_enabled = !local.is_sandbox
+
+  attributes = [
+    {
+      name = "ID"
+      type = "S"
+    },
```
Comment (chrisbloe-nhse) [inline infrastructure/dynamo_db.tf:57]: trailing comma

File: infrastructure/lambda-edge-presign.tf [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
+    aws_iam_policy.ssm_policy_oidc.arn,
+    module.auth_state_dynamodb_table.dynamodb_policy,
+    module.ndr-app-config.app_config_policy_arn
+  ]
+  providers = {
+    aws = aws.us_east_1
+  }
+  current_account_id = data.aws_caller_identity.current.account_id
+  bucket_name        = "${terraform.workspace}_${var.lloyd_george_bucket_name}"
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-edge-presign.tf:58]: `module.ndr-lloyd-george-store.bucket_id`

File: infrastructure/lambda-edge-presign.tf [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  table_name         = "${terraform.workspace}_${var.cloudfront_edge_table_name}"
```
Comment (chrisbloe-nhse) [inline infrastructure/lambda-edge-presign.tf:59]: `module.cloudfront_edge_dynamodb_table.table_name`

Extra needed within module (`national-document-repository-infrastructure/infrastructure/modules/dynamo_db/variable.tf`):
```
output "table_name" {
  value = aws_dynamodb_table.ndr_dynamodb_table.id
}
```

File: infrastructure/buckets.tf [hunk 5]
```diff
@@ -62,6 +63,15 @@ module "ndr-lloyd-george-store" {
   ]
 }
 
+module "cloudfront-distribution-lg" {
```
Comment (chrisbloe-nhse) [inline infrastructure/buckets.tf:66]: This should be in its own `cloudfront.tf` file (not in with S3 resources)

File: infrastructure/main.tf [hunk 6]
```diff
@@ -18,6 +18,10 @@ provider "aws" {
   region = "eu-west-2"
 }
 
+provider "aws" {
```
Comment (chrisbloe-nhse) [inline infrastructure/main.tf:21]: The `aws` provider version would be better locked to `5` e.g. `version ="~> 5.0"`

File: infrastructure/modules/lambda_edge/main.tf [hunk 7]
```diff
@@ -0,0 +1,92 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">=5.11"
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/lambda_edge/main.tf:5]: The `aws` provider version would be better locked to `5` e.g. `version ="~> 5.0"`

File: infrastructure/modules/cloudfront/variable.tf [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  description = "Bucket ID to assign CloudFront distribution to"
+}
+
+variable "qualifed_arn" {
+  type        = string
+  description = "Lambda@Edge function association"
+}
+
+
+output "cloudfront_url" {
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/cloudfront/variable.tf:17]: Please could these go in an `output.tf` file?

File: infrastructure/modules/cloudfront/main.tf [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+resource "aws_cloudfront_distribution" "distribution" {
+  origin {
+    domain_name              = var.bucket_domain_name
+    origin_id                = var.bucket_id
+    origin_access_control_id = aws_cloudfront_origin_access_control.cloudfront_s3_oac.id
+  }
+  enabled         = true
+  is_ipv6_enabled = true
+  default_cache_behavior {
+    allowed_methods  = ["GET", "HEAD"]
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/cloudfront/main.tf:18]: `,"OPTIONS"`?

File: infrastructure/modules/cloudfront/main.tf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    cached_methods   = ["GET", "HEAD"]
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/cloudfront/main.tf:19]: Do we want `GET` to be cached?

File: infrastructure/modules/cloudfront/main.tf [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    target_origin_id = var.bucket_id
+    # viewer_protocol_policy = "redirect-to-https"
+
+    forwarded_values {
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/cloudfront/main.tf:23]: This is [deprecated](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudfront_distribution#forwarded_values) :/.

File: infrastructure/modules/cloudfront/main.tf [hunk 12]
```diff
@@ -0,0 +1,50 @@
+resource "aws_cloudfront_origin_access_control" "cloudfront_s3_oac" {
+  name                              = "CloudFront S3 OAC"
+  description                       = "Cloud Front S3 OAC"
+  origin_access_control_origin_type = "s3"
+  signing_behavior                  = "always"
+  signing_protocol                  = "sigv4"
+}
+
+resource "aws_cloudfront_distribution" "distribution" {
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/cloudfront/main.tf:9]: Missing `logging_config` block

File: infrastructure/modules/cloudfront/main.tf [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  origin {
+    domain_name              = var.bucket_domain_name
+    origin_id                = var.bucket_id
+    origin_access_control_id = aws_cloudfront_origin_access_control.cloudfront_s3_oac.id
+  }
+  enabled         = true
+  is_ipv6_enabled = true
+  default_cache_behavior {
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/cloudfront/main.tf:17]: Needs a `cache_policy_id = aws_cloudfront_cache_policy.nocache.id`

Then something like:
```
resource "aws_cloudfront_cache_policy" "nocache" {
  name        = "nocache_policy"
  default_ttl = 0
  max_ttl     = 0
  min_ttl     = 0

  parameters_in_cache_key_and_forwarded_to_origin {
    cookies_config {
      cookie_behavior = "none"
    }
    headers_config {
      header_behavior = "none"
    }
    query_strings_config {
      query_string_behavior = "none"
    }
  }
}
```

File: infrastructure/modules/lambda/variable.tf [hunk 14]
```diff
@@ -66,10 +66,21 @@ variable "reserved_concurrent_executions" {
   default     = -1
 }
 
+variable "is_edge" {
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/lambda/variable.tf:69]: Is this used??

File: infrastructure/modules/lambda_edge/main.tf [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+}
+
+resource "aws_lambda_function" "lambda" {
+  provider = aws # Alternative AWS provider for Lambda@Edge region
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/lambda_edge/main.tf:11]: Does this point at Virginia?

File: infrastructure/modules/s3/variable.tf [hunk 16]
```diff
@@ -31,6 +33,18 @@ variable "owner" {
   type = string
 }
 
+variable "cloudfront_arn" {
+  type        = string
+  default     = "null"
+  description = "Cloudfront Distribution ARN association and policy toggles"
+}
+
+variable "cloudfront_enabled" {
```
Comment (chrisbloe-nhse) [inline infrastructure/modules/s3/variable.tf:42]: Is this used?

File: infrastructure/main.tf (hunk 6 above)
Comment (RioKnightleyNHS) [reply infrastructure/main.tf:21]: depreciated in place of 'required_providers' in the global terraform block, which makes sure all imported aws providers adhere to one version

File: infrastructure/modules/lambda_edge/main.tf (hunk 7 above)
Comment (RioKnightleyNHS) [reply infrastructure/modules/lambda_edge/main.tf:5]: Agreed but I think "~> 5.11.0" could be better else we would be rolling back our provider tf versioning

File: infrastructure/modules/cloudfront/main.tf (hunk 12 above)
Comment (RioKnightleyNHS) [reply infrastructure/modules/cloudfront/main.tf:9]: are we prepared to create a new bucket for cloudfront logs ?

File: infrastructure/lambda-edge-presign.tf [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+  delivery_policy = jsonencode({
+    "Version" : "2012-10-17",
+    "Statement" : [
+      {
+        "Effect" : "Allow",
+        "Principal" : {
+          "Service" : "cloudwatch.amazonaws.com"
+        },
+        "Action" : [
+          "SNS:Publish",
```
Comment (chrisbloe) [inline infrastructure/lambda-edge-presign.tf:30]: Trailing comma

File: infrastructure/modules/s3/variable.tf [hunk 18]
```diff
@@ -31,19 +33,14 @@ variable "owner" {
...
-output "s3_list_object_policy" {
-  value = aws_iam_policy.s3_list_object_policy.arn
-}
-
-output "bucket_id" {
-  value = aws_s3_bucket.bucket.id
+variable "cloudfront_arn" {
+  type        = string
+  default     = "null"
+  description = "Cloudfront Distribution ARN association and policy toggles"
```
Comment (chrisbloe) [inline infrastructure/modules/s3/variable.tf:39]: `CloudFront`

File: infrastructure/modules/s3/variable.tf [hunk 19]
```diff
@@ -31,19 +33,14 @@ variable "owner" {
...
 }
 
-output "bucket_arn" {
-  value = aws_s3_bucket.bucket.arn
+variable "cloudfront_enabled" {
+  type        = bool
+  default     = false
+  description = "Enables the correct policy config for cloudfront s3"
```
Comment (chrisbloe) [inline infrastructure/modules/s3/variable.tf:45]: `CloudFront S3`

File: infrastructure/variable.tf [hunk 20]
```diff
@@ -60,6 +60,18 @@ variable "lloyd_george_dynamodb_table_name" {
   default     = "LloydGeorgeReferenceMetadata"
 }
 
+variable "cloudfront_edge_table_name" {
+  type        = string
+  description = "The name of dynamodb table to store the metadata of Lloyd George documents"
```
Comment (chrisbloe) [inline infrastructure/variable.tf:65]: This is the same description as the variable below?!

File: infrastructure/modules/app_config/configurations/2024-05-21-default.json [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+            "enabled": "false"
+        },
+        "uploadLambdaEnabled": {
+            "enabled": "false"
+        },
+        "uploadArfWorkflowEnabled": {
+            "enabled": "false"
+        },
+        "useSmartcardAuth": {
+            "enabled": "true"
```
Comment (abbas-khan10) [inline infrastructure/modules/app_config/configurations/2024-05-21-default.json:33]: What's the reason for this new file?

File: infrastructure/modules/lambda_edge/main.tf [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+  provider = aws
+  publish  = true
+
+  filename                       = data.archive_file.lambda.output_path
+  function_name                  = "${terraform.workspace}_${var.name}"
+  role                           = aws_iam_role.lambda_exec_role.arn
+  handler                        = var.handler
+  source_code_hash               = data.archive_file.lambda.output_base64sha256
+  runtime                        = "python3.11"
+  timeout                        = 5
```
[the remark this answers is not in the document]
Comment (RioKnightleyNHS) [reply infrastructure/modules/lambda_edge/main.tf:20]: it's the maximum allowed for an edge lambda

File: infrastructure/modules/s3/variable.tf [hunk 23]
```diff
@@ -3,6 +3,8 @@ variable "bucket_name" {
   type        = string
 }
 
+
+
```
Comment (abbas-khan10) [inline infrastructure/modules/s3/variable.tf:7]: 3 break lines?

File: scripts/remove_edge_associations.py [hunk 24]
```diff
@@ -0,0 +1,85 @@
+import boto3
+from botocore.exceptions import ClientError
+import time
+
+def log(message):
+    print(message, flush=True)
+
+def detach_lambda_edge_associations(distribution_id):
```
Comment (abbas-khan10) [inline scripts/remove_edge_associations.py:8]: Could we add in expected types in here to improve readability of the script? There's a few missing

File: scripts/remove_edge_associations.py [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    try:
+        client = boto3.client('cloudfront')
+        
+        response = client.get_distribution_config(Id=distribution_id)
+        config = response['DistributionConfig']
+        etag = response['ETag']
+
+        behaviors = []
```
Comment (abbas-khan10) [inline scripts/remove_edge_associations.py:16]: Missing type e.g. `behaviours: list[str] = []` just so it's clear what we expect this array to hold

File: scripts/remove_edge_associations.py [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+    lambda_function_name = os.getenv("LAMBDA_FUNCTION_NAME")
+
+    if not distribution_id:
+        raise ValueError("The DISTRIBUTION_ID environment variable is not set.")
+    if not lambda_function_name:
+        raise ValueError("The LAMBDA_FUNCTION_NAME environment variable is not set.")
+
+    detach_lambda_edge_associations(distribution_id)  
+    
+    # if delete_lambda_function(lambda_function_name, max_retries=30, initial_wait_time=30):
```
Comment (abbas-khan10) [inline scripts/remove_edge_associations.py]: leftover code?

**Review by abbas-khan10 — APPROVED** [review]

**Pull request merged**